### ISSUE TITLE
super(type, instance) in classmethod is not bound

### ISSUE THREAD

**Issue by hrmorley34** [microsoft/pylance-release#2230]
<h3> Environment data </h3>

-   Language Server version: 2022.1.0
-   OS and version: win32 x64
-   Python version: 3.10.1
-   python.analysis.indexing: undefined
-   python.analysis.typeCheckingMode: strict

<h3> Expected behaviour </h3>

`super(type, instance)` should always give access to *bound* methods.

<h3> Actual behaviour </h3>

`super(type, instance)`, when invoked within a `classmethod` or `staticmethod`, is given a type of `Type[Class]` rather than `Class`, and therefore methods of it are considered unbound and require one extra positional argument.

<h3> Logs </h3>

<details>
<summary>Python Language Server Log</summary>

```
[Info  - 23:41:19] Pylance language server 2022.1.0 (pyright 5829dcbb) starting
[Info  - 23:41:19] Server root directory: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist
[Info  - 23:41:19] No configuration file found.
[Info  - 23:41:19] No pyproject.toml file found.
[Info  - 23:41:19] Setting pythonPath for service "(workspace)": "C:\Users\henry\AppData\Local\Programs\Python\Python310\python.exe"
[Warn  - 23:41:19] stubPath (workspace)\typings is not a valid directory.
[Info  - 23:41:19] Assuming Python version 3.10
[Info  - 23:41:19] Assuming Python platform Windows
Search paths for (workspace)
  c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib
  (workspace)
  (workspace)\typings
  c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stubs\...
  c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\bundled\stubs
  c:\Users\henry\AppData\Local\Programs\Python\Python310\DLLs
  c:\Users\henry\AppData\Local\Programs\Python\Python310\Lib
  c:\Users\henry\AppData\Local\Programs\Python\Python310
  c:\Users\henry\AppData\Local\Programs\Python\Python310\Lib\site-packages
  (...)
  c:\Users\henry\AppData\Local\Programs\Python\Python310\Lib\site-packages\win32
  c:\Users\henry\AppData\Local\Programs\Python\Python310\Lib\site-packages\win32\lib
  c:\Users\henry\AppData\Local\Programs\Python\Python310\Lib\site-packages\pythonwin
[Info  - 23:41:20] Searching for source files
[Info  - 23:41:20] Found 9 source files
[Info  - 23:41:20] Background analysis(1) root directory: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist
[Info  - 23:41:20] Background analysis(1) started
Background analysis message: setConfigOptions
Background analysis message: setImportResolver
Background analysis message: ensurePartialStubPackages
[FG] parsing: (workspace)\example.py (48ms)
[FG] parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 5ms] (188ms)
[FG] binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\builtins.pyi (78ms)
[FG] binding: (workspace)\example.py (2ms)
Background analysis message: setTrackedFiles
Background analysis message: markAllFilesDirty
Background analysis message: setFileOpened
Background analysis message: analyze
IntelliCode model c:\Users\henry\.vscode\extensions\visualstudioexptteam.vscodeintellicode-1.2.15\cache\E61945A9A512ED5E1A3EE3F1A2365B88F8FE_E4E9EADA96734F01970E616FAB2FAC19
[BG(1)] analyzing: (workspace)\example.py ...
[BG(1)]   parsing: (workspace)\example.py (79ms)
[BG(1)]   parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 6ms] (327ms)
[BG(1)]   binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\builtins.pyi (131ms)
[BG(1)]   binding: (workspace)\example.py (3ms)
[BG(1)]   checking: (workspace)\example.py ...
[BG(1)]     parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\typing_extensions.pyi [fs read 6ms] (40ms)
[BG(1)]     binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\typing_extensions.pyi (3ms)
[BG(1)]     parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\typing.pyi [fs read 1ms] (98ms)
[BG(1)]     binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\typing.pyi (34ms)
[BG(1)]     parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi [fs read 1ms] (30ms)
[BG(1)]     binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi (4ms)
[BG(1)]     parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\types.pyi [fs read 1ms] (29ms)
[BG(1)]     binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\types.pyi (22ms)
[BG(1)]     parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\abc.pyi [fs read 1ms] (2ms)
[BG(1)]     binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\abc.pyi (1ms)
[BG(1)]     parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\traceback.pyi [fs read 1ms] (11ms)
[BG(1)]     binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\traceback.pyi (4ms)
[BG(1)]   checking: (workspace)\example.py (494ms)
[BG(1)] analyzing: (workspace)\example.py (1041ms)
Background analysis message: getSemanticTokens full
[BG(1)] getSemanticTokens full at (workspace)\example.py (42ms)
Background analysis message: getSemanticTokens range
[BG(1)] getSemanticTokens range 39:0 - 90:0 at (workspace)\example.py (41ms)
Background analysis message: resumeAnalysis
Loading ONNX runtime...
Loaded ONNX runtime. Creating IntelliCode session...
Created IntelliCode session.
Initialize deeplearning succeeded
[FG] parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\typing_extensions.pyi [fs read 2ms] (18ms)
[FG] binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\typing_extensions.pyi (6ms)
[FG] parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\typing.pyi [fs read 3ms] (56ms)
[FG] binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\typing.pyi (12ms)
[FG] parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi [fs read 2ms] (15ms)
[FG] binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi (3ms)
[FG] parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\types.pyi [fs read 1ms] (25ms)
[FG] binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\types.pyi (5ms)
[FG] parsing: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\abc.pyi [fs read 1ms] (3ms)
[FG] binding: c:\Users\henry\.vscode\extensions\ms-python.vscode-pylance-2022.1.0\dist\typeshed-fallback\stdlib\abc.pyi (1ms)
[FG] parsing: C:\Users\henry\AppData\Local\Temp\pyright-15380-4yyQnIjee5jL\builtins-15380-GMgPIxJ7rLBH-.py [fs read 17ms] (185ms)
[FG] binding: C:\Users\henry\AppData\Local\Temp\pyright-15380-4yyQnIjee5jL\builtins-15380-GMgPIxJ7rLBH-.py (45ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
(etc.)
```
</details>

<h3> Code Snippet / Additional information </h3>

```python
from traceback import print_exc
from typing import TYPE_CHECKING, Any


if not TYPE_CHECKING:
    def reveal_type(obj: Any) -> None:
        pass


class BaseClass:
    def my_method(self, value: int) -> int:
        return value


class SubClass(BaseClass):
    def method_plain_super(self, value: int) -> int:
        reveal_type(super())  # => BaseClass
        return super().my_method(value)

    # correctly accepted
    def method_super(self, value: int) -> int:
        reveal_type(super(__class__, self))  # => BaseClass
        return super(__class__, self).my_method(value)

    # correctly erroneous
    def method_super_extra_arg(self, value: int) -> int:
        reveal_type(super(__class__, self))  # => BaseClass
        # Return type is unknown Pylance(reportUnknownVariableType)
        # Expected 1 positional argument Pylance(reportGeneralTypeIssues)
        return super(__class__, self).my_method(self, value)

    # incorrectly erroneous
    @classmethod
    def classmethod_super(cls, value: int) -> int:
        self = cls()
        reveal_type(super(__class__, self))  # => Type[BaseClass]
        # Argument missing for parameter "value" Pylance(reportGeneralTypeIssues)
        # Return type is unknown Pylance(reportUnknownVariableType)
        return super(__class__, self).my_method(value)

    # incorrectly accepted
    @classmethod
    def classmethod_super_extra_arg(cls, value: int) -> int:
        self = cls()
        reveal_type(super(__class__, self))  # => Type[BaseClass]
        return super(__class__, self).my_method(self, value)

    # same problem with staticmethod as with classmethod

    @staticmethod
    def staticmethod_super(value: int) -> int:
        self = SubClass()
        reveal_type(super(__class__, self))  # => Type[BaseClass]
        # Argument missing for parameter "value" Pylance(reportGeneralTypeIssues)
        # Return type is unknown Pylance(reportUnknownVariableType)
        return super(__class__, self).my_method(value)

    @staticmethod
    def staticmethod_super_extra_arg(value: int) -> int:
        self = SubClass()
        reveal_type(super(__class__, self))  # => Type[BaseClass]
        # accepted
        return super(__class__, self).my_method(self, value)


print(SubClass().method_plain_super(0))  # => 0
print(SubClass().method_super(0))  # => 0

try:
    # TypeError: Test.method1() takes 2 positional arguments but 3 were given
    print(SubClass().method_super_extra_arg(0))
except:
    print_exc()

print(SubClass().classmethod_super(0))  # => 0

try:
    # TypeError: Test.method1() takes 2 positional arguments but 3 were given
    print(SubClass().classmethod_super_extra_arg(0))
except:
    print_exc()

print(SubClass().staticmethod_super(0))  # => 0

try:
    # TypeError: Test.method1() takes 2 positional arguments but 3 were given
    print(SubClass().staticmethod_super_extra_arg(0))
except:
    print_exc()
```

**Comment by erictraut** [comment]
Thanks for the bug report. I agree this is a problem. I'll work on a fix.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2022.1.1, which we've just released. You can find the changelog here:  [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202211-12-january-2022)
